cpp/src/arrow/dataset/file_skyhook.h

```diff
@@ -0,0 +1,275 @@
+      : Connection(),
+        ctx(ctx),
+        rados(new RadosWrapper()),
+        ioCtx(new IoCtxWrapper()),
+        connected(false) {}
+
+  ~RadosConnection();
+
+  /// \brief Connect to the Rados cluster.
+  /// \return Status.
+  Status Connect();
+
+  /// \brief Shutdown the connection to the Rados cluster.
+  /// \return Status.
+  Status Shutdown();
+
+  RadosConnectionCtx ctx;
+  RadosInterface* rados;
+  IoCtxInterface* ioCtx;
+  bool connected;
+};
+
+}  // namespace connection
+
+/// \class SkyhookDirectObjectAccess
+/// \brief Interface for translating the name of a file in CephFS to its
+/// corresponding object ID in RADOS assuming 1:1 mapping between a file
+/// and its underlying object.
+class ARROW_DS_EXPORT SkyhookDirectObjectAccess {
+ public:
+  explicit SkyhookDirectObjectAccess(
+      const std::shared_ptr<connection::RadosConnection>& connection)
+      : connection_(std::move(connection)) {}
+
+  /// \brief Executes the POSIX stat call on a file.
+  /// \param[in] path Path of the file.
+  /// \param[out] st Refernce to the struct object to store the result.
+  /// \return Status.
+  Status Stat(const std::string& path, struct stat& st) {
+    struct stat file_st;
+    if (stat(path.c_str(), &file_st) < 0)
+      return Status::Invalid("stat returned non-zero exit code.");
+    st = file_st;
+    return Status::OK();
+  }
+
+  // Helper function to convert Inode to ObjectID because Rados calls work with
+  // ObjectIDs.
+  std::string ConvertFileInodeToObjectID(uint64_t inode) {
+    std::stringstream ss;
+    ss << std::hex << inode;
+    std::string oid(ss.str() + ".00000000");
+    return oid;
+  }
+
+  /// \brief Executes query on the librados node. It uses the librados::exec API to
+  /// perform queries on the storage node and stores the result in the output bufferlist.
+  /// \param[in] inode inode of the file.
+  /// \param[in] fn The function to be executed by the librados::exec call.
+  /// \param[in] in The input bufferlist.
+  /// \param[out] out The output bufferlist.
+  /// \return Status.
+  Status Exec(uint64_t inode, const std::string& fn, ceph::bufferlist& in,
+              ceph::bufferlist& out) {
+    std::string oid = ConvertFileInodeToObjectID(inode);
+    int e = connection_->ioCtx->exec(oid.c_str(), connection_->ctx.cls_name.c_str(),
+                                     fn.c_str(), in, out);
+    if (e == SCAN_ERR_CODE) return Status::Invalid(SCAN_ERR_MSG);
+    if (e == SCAN_REQ_DESER_ERR_CODE) return Status::Invalid(SCAN_REQ_DESER_ERR_MSG);
+    if (e == SCAN_RES_SER_ERR_CODE) return Status::Invalid(SCAN_RES_SER_ERR_MSG);
+    return Status::OK();
+  }
+
+ protected:
+  std::shared_ptr<connection::RadosConnection> connection_;
+};
+
+/// \class SkyhookFileFormat
+/// \brief A ParquetFileFormat implementation that offloads the fragment
+/// scan operations to the Ceph OSDs
+class ARROW_DS_EXPORT SkyhookFileFormat : public ParquetFileFormat {
```

Comment:
I think this is not the interface you want to implement for skyhook.

A `FileFormat`'s fragments should be able to wrap files in any file system- or even a buffer in memory instead of a file. However IIUC `SkyhookFileFormat` always assumes we can do direct object access, which assumes we have a path and a file system which supports direct object access.

Instead I think this should be exposed as a `Fragment` subclass (**not** a `FileFragment`), or maybe even an implementation of `Scanner` since it overrides client-side filtering and projection.

Reply:
Thank you, @bkietz -- I agree with you that `FileFormat` is not the ideal place. I think we proposed something similar to what you suggest in @ivotron's [pull request](https://github.com/apache/arrow/pull/8647) (see also [mailing list thread](https://lists.apache.org/thread.html/r2a5a693967213b7c6bb49015194ca16afc4d20047805d0e069c2e45c%40%3Cdev.arrow.apache.org%3E)). However he ended up closing that pull request because he realized we would be re-implementing a lot of file system functionality. I've just talked to @ivotron about this and he pointed out that  the fact that we are getting bounced between these two approaches shows that the notion of offloading computation to the storage layer is a more fundamental issue than a specialization of`Scanner` or `FileFormat`. I hope that the `SkyhookFileFormat` can be merged to provide a use case for providing a more general solution to offloading in Arrow.

Reply:
> the notion of offloading computation to the storage layer is a more fundamental issue than a specialization of Scanner or FileFormat

This is likely true.  It would be good to start brainstorming what sorts of concepts we are missing.  The mailing list could be a good place for discussion & brainstorming or future PRs could be a good place for more concrete concepts/proposals.

At the moment we have (I'll add that these APIs have evolved considerably since the start of the year):

 * **FileFormat** - A file format converts a stream of bytes (regardless of source) into a stream of record batches.
 * **Scanner** - A scanner converts a dataset (list of fragments) into a stream of record batches.  The scanner also applies filtering & projection.
 *  **Fragment** - An independently scannable unit
 *  **FileFragment** - An independently scannable unit backed by a file, relies on `FileFormat`.

The current scanner is also a fragment scheduler.  Given a list of fragments it selects N fragments to start scanning and merges (and sometimes sequences) their results.

Arrow itself is also in flux with these interfaces.  The scanner is being split into a scan node which does fragment scheduling and the filtering/projection pieces are being pulled into their own nodes.

The downside of skyhook being a scanner that I can see is that skyhook doesn't do any fragment scheduling today, and it would have to reinvent that.  On the other hand, maybe skyhook knows more information that would help it do this scheduling.  For example, Arrow just sort of picks "8 fragments at a time" is probably good enough (kDefaultFragmentReadahead) which is overkill for HDD, about right for SDD and S3, and underkill for certain EC2/S3 configurations.  I have no idea what is appropriate for skyhook (I assume it would be based on the # of nodes in the ceph cluster and maybe the file distribution).  This particular fact wouldn't necessarily require an additional scanner implementation but maybe some other fact does (e.g. maybe there is a certain ordering in which fragments should be scheduled).  I can understand why one wouldn't want Skyhook to be a Scanner.

The downside of skyhook being a file format is that, if it isn't the right fit, it starts to break apart the abstraction.  For example, we have to introduce `FileFormat::dataset_schema`, there are potential Liskov violations which could mean the code is more brittle (i.e. we might start using FileFormat differently or elsewhere and the Rados implementation doesn't fit the new usage).

Perhaps Skyhook is a good fit for `Fragment`.  This would allow Skyhook to take advantage of the Scanner's scheduling but give it more flexibility.  Adding `handles_filtering` and `handles_projection` properties to a fragment seems appropriate too.  Or, Skyhook could attach a "guarantee" to the returned batch (this concept isn't in scanner yet but it is emerging in the exec plan abstractions) so that future filtering mechanisms would skip it because the guarantee shows it is already filtered.

I'm also pretty amenable to the logic of "we don't need to get it perfect now" and we can get something working and iterate on it in the future.  I appreciate the effort to break things into smaller PRs.